### PR TITLE
More logic changes to error reporting, cleanup

### PR DESCRIPTION
**Pull request by agjohnson**
* Alter project.task logic and include multiple points of error reporting during
  build task
* Build task returns true or false, not project data, which is unused anyways
* New pattern for error messages
* Localize error messages?
* Make more specific error messages for repository error and project error
* Remove unnecessary git/hg/bzr/svn exit codes and superfluous command data from
  user errors
* Add handling of git prompt for git 2.3+
* Add errors for docker connection
* Handle reporting on not able to connect to docker

### CODE REVIEW

**Comment by agjohnson** [comment]
Marked WIP, because:

- [ ] Regression in reporting build finished in `run_setup()`
- [ ] Not unittested yet
- [ ] Need to play with exception translation pattern -- ultimately not worried about this yet, it will be an eventual improvement however.
- [ ] Need to do more to catch an API exception with docker on memory issues and report this
- [ ] Should actually write unittests around all of this

**Comment by agjohnson** [comment]
Also, I'm thinking that we need to move the vcs execution into a build environment, the error reporting with git stderr is sloppy, but there's no reason we can't just show this step to users. This will be a separate ticket.

File: readthedocs/doc_builder/environments.py [hunk 1]
```diff
@@ -550,15 +553,17 @@ def __exit__(self, exc_type, exc_value, tb):
         try:
             log.info('Removing container %s', self.container_id)
             client.remove_container(self.container_id)
-        except DockerAPIError:
+        # Catch direct failures from Docker API, but also requests exceptions
+        # with the HTTP request
+        except (DockerAPIError, ConnectionError):
             log.error(LOG_TEMPLATE
```
Comment (humitos) [inline readthedocs/doc_builder/environments.py]: isn't it better here to use `log.exception` ?

File: readthedocs/projects/exceptions.py [hunk 2]
```diff
@@ -1,15 +1,46 @@
...
+from django.conf import settings
+from django.utils.translation import ugettext_noop as _
 
-class ProjectImportError (Exception):
+from readthedocs.doc_builder.exceptions import BuildEnvironmentError
 
-    """Failure to import a project from a repository."""
 
-    pass
+class ProjectConfigurationError(BuildEnvironmentError):
```
Comment (humitos) [inline readthedocs/projects/exceptions.py]: I think this class is missing the `get_default_message` method.

File: readthedocs/vcs_support/backends/git.py [hunk 3]
```diff
@@ -58,10 +59,8 @@ def repo_exists(self):
     def fetch(self):
         code, _, err = self.run('git', 'fetch', '--tags', '--prune')
```
Comment (humitos) [inline readthedocs/vcs_support/backends/git.py]: `err` could be `_` now that's not used

File: readthedocs/vcs_support/backends/git.py [hunk 4]
```diff
@@ -79,16 +78,7 @@ def clone(self):
         code, _, err = self.run('git', 'clone', '--recursive', '--quiet',
```
Comment (humitos) [inline readthedocs/vcs_support/backends/git.py]: `err` could be `_` now that's not used

File: readthedocs/vcs_support/backends/git.py [hunk 5]
```diff
@@ -79,16 +78,7 @@ def clone(self):
...
-                (
-                    "Failed to get code from '{url}' (git clone): {exit}\n\n"
-                    "git clone error output: {sterr}"
-                ).format(
-                    url=self.repo_url,
-                    exit=code,
-                    sterr=err
-                )
-            )
+            raise RepositoryError
```
Comment (humitos) [inline readthedocs/vcs_support/backends/git.py]: I think this is too much simpler that what we had, but what happen if there is another problem different than a private repository. I mean, the error message is very generic and we are not checking a specific error code:

For example:

```
[humitos@julia:tmp]$ GIT_TERMINAL_PROMPT=0 git clone https://github.com/readthedocs/readthedocs-corporate
Clonando en 'readthedocs-corporate'...
fatal: could not read Username for 'https://github.com': terminal prompts disabled
[humitos@julia:tmp]$  echo $?
128
```

Ha, it doesn't seem to have a different error code:
```
[humitos@julia:tmp]$ LANG=C git clone git@github.com:rtfd/readthedocs
Cloning into 'readthedocs'...
ERROR: Repository not found.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
[humitos@julia:tmp]$ echo $?
128
```

Anyway, I think this is not really important if we don't have history/logs failing at these points because something different than a private repository.

File: readthedocs/projects/tasks.py [hunk 6]
```diff
@@ -127,24 +126,40 @@ def run(self, pk, version_pk=None, build_pk=None, record=True,
             self.config = None
 
             setup_successful = self.run_setup(record=record)
-            if setup_successful:
-                self.run_build(record=record, docker=docker)
+            if not setup_successful:
```
Comment (humitos) [inline readthedocs/projects/tasks.py]: I _think_ there is something missing here. If the build fails because of the setup at this point: https://github.com/rtfd/readthedocs.org/pull/3310/files#diff-b9399e1d3499066c5564f98a620e8881R198

we are not raising any exception, so `run_setup` method just returns `False` and the task return `False` but nobody is updating the `self.setup_env.build['error']` as we do when unhandled exception is catched at https://github.com/rtfd/readthedocs.org/pull/3310/files#diff-b9399e1d3499066c5564f98a620e8881R137 or https://github.com/rtfd/readthedocs.org/pull/3310/files#diff-b9399e1d3499066c5564f98a620e8881R154

Also, I think that's better to move this line https://github.com/rtfd/readthedocs.org/pull/3310/files#diff-b9399e1d3499066c5564f98a620e8881R207 inside this if, so we try to update the build state in not so many places

File: readthedocs/vcs_support/backends/git.py (hunk 5 above)
Comment (agjohnson) [reply readthedocs/vcs_support/backends/git.py]: I agree, but this would be out of scope for this work, and would be best served by surfacing the commands we execute to users. Currently, the checkout step is hidden because it uses it's own shell execution. This could be easily moved into a BuildCommand executed through the setup environment. This way, users can actually see the output in a meaningful way.

Comment (humitos) [reply readthedocs/vcs_support/backends/git.py]: I like your proposal of moving it inside a BuildCommand and I agree it could be done in another PR.

File: readthedocs/projects/exceptions.py (hunk 2 above)
Comment (agjohnson) [reply readthedocs/projects/exceptions.py]: It should be inheriting this from `BuildEnvironmentException`

**Comment by ericholscher** [comment]
Feel like this is a biggish change, which might conflict over time -- should try and get tests passing and work done, and then I'll review it.

**Comment by agjohnson** [comment]
Test failure was unrelated.

File: readthedocs/projects/tasks.py (hunk 6 above)
Comment (agjohnson) [reply readthedocs/projects/tasks.py]: Links above have disconnected here, so hard to follow what's happening, but i think i understand this feedback.

So update of the build now always happens on build env `__exit__`, so we can just return false from the task.

In the case that another exception is raised outside the env, we still run `update_build` from the exception.

**Comment by agjohnson** [comment]
Unittests around doc build environments have been made working, and more have been added.

a9c2a22 is lint clean up, review that separately.

**Comment by ericholscher** [comment]
How do I review it seperately?

**Comment by humitos** [comment]
You can select a range of commits to review from the top left dropdown: https://github.com/rtfd/readthedocs.org/pull/3310/files/0fd836669a2a3f87866634c534d3a32c7dff462b

(just the 2 commits to review with logic)

File: readthedocs/doc_builder/backends/sphinx.py [hunk 7]
```diff
@@ -121,16 +121,22 @@ def append_conf(self, **__):
...
             self._write_config(master_doc=master_doc)
 
         try:
             outfile_path = self.project.conf_file(self.version.slug)
             outfile = codecs.open(outfile_path, encoding='utf-8', mode='a')
-        except (ProjectImportError, IOError):
+        except (ProjectConfigurationError, IOError):
             trace = sys.exc_info()[2]
-            six.reraise(ProjectImportError('Conf file not found'), None, trace)
+            six.reraise(
```
Comment (humitos) [inline readthedocs/doc_builder/backends/sphinx.py]: Why do we need this `six.reraise`? It's not the same than just `raise ProjectImportError(...)`?

**Comment by ericholscher** [comment]
@humitos thanks!

File: readthedocs/doc_builder/environments.py [hunk 8]
```diff
@@ -294,7 +297,7 @@ def __enter__(self):
 
     def __exit__(self, exc_type, exc_value, tb):
         ret = self.handle_exception(exc_type, exc_value, tb)
-        self.build['state'] = BUILD_STATE_FINISHED
+        self.update_build(BUILD_STATE_FINISHED)
```
Comment (ericholscher) [inline readthedocs/doc_builder/environments.py]: Won't this post that status to the API? Seems like we don't want to publish here, but only below if `finalize` is `False`.

File: readthedocs/doc_builder/environments.py [hunk 9]
```diff
@@ -521,8 +525,15 @@ def __enter__(self):
...
                     client.remove_container(self.container_id)
-        except DockerAPIError:
+        except (DockerAPIError, ConnectionError):
+            # If there is an exception here, we swallow the exception as this
+            # was just during a sanity check anyways.
             pass
+        except BuildEnvironmentError:
+            # There may have been a problem connecting to Docker altogether, or
+            # some other handled exception here.
+            self.__exit__(*sys.exc_info())
```
Comment (ericholscher) [inline readthedocs/doc_builder/environments.py]: Will this not automatically call `__exit__`?

File: readthedocs/doc_builder/environments.py [hunk 10]
```diff
@@ -275,15 +276,17 @@ class BuildEnvironment(object):
     :param build: Build instance
     :param record: Record status of build object
     :param environment: shell environment variables
+    :param finalize: finalize the build by setting a finished state on exit
```
Comment (humitos) [inline readthedocs/doc_builder/environments.py]: this sentence is `True` _only_ if the `update_build` method is called with the `BUILD_STATE_FINISHED` status, otherwise that method will just update the build, but not finalize it

File: readthedocs/doc_builder/environments.py [hunk 11]
```diff
@@ -537,28 +548,40 @@ def __enter__(self):
...
-            log.error(LOG_TEMPLATE
-                      .format(
-                          project=self.project.slug,
-                          version=self.version.slug,
-                          msg="Couldn't remove container"),
-                      exc_info=True)
-        self.container = None
-        self.build['state'] = BUILD_STATE_FINISHED
+        ret = self.handle_exception(exc_type, exc_value, tb)
+        self.update_build(BUILD_STATE_FINISHED)
```
Comment (ericholscher) [inline readthedocs/doc_builder/environments.py]: Feels like we're posting this to the API a lot more with these changes -- should we not respect `finalize` everywhere we set build state to finished?

File: readthedocs/doc_builder/environments.py [hunk 12]
```diff
@@ -537,28 +548,40 @@ def __enter__(self):
...
+            # Update buildenv state given any container error states first
+            self.update_build_from_container_state()
 
-        # Update buildenv state given any container error states first
-        self.update_build_from_container_state()
+            client = self.get_client()
+            try:
+                client.kill(self.container_id)
+            except DockerAPIError:
+                pass
```
Comment (humitos) [inline readthedocs/doc_builder/environments.py]: what about logging something here (DEBUG level, maybe)?

File: readthedocs/doc_builder/environments.py [hunk 13]
```diff
@@ -537,28 +548,40 @@ def __enter__(self):
...
+            try:
+                log.info('Removing container %s', self.container_id)
+                client.remove_container(self.container_id)
+            # Catch direct failures from Docker API, but also requests exceptions
+            # with the HTTP request. These should not
```
Comment (humitos) [inline readthedocs/doc_builder/environments.py]: "These should not" ?

File: readthedocs/projects/exceptions.py (hunk 2 above)
Comment (ericholscher) [inline readthedocs/projects/exceptions.py]: It seems we have a `BuildEnvironmentException` and a `BuildEnvironmentError`, what is the difference? Seems confusing..

File: readthedocs/projects/exceptions.py [hunk 14]
```diff
@@ -1,15 +1,46 @@
...
+
+
+class RepositoryError(BuildEnvironmentError):
+
+    """Failure during repository operation."""
+
+    PRIVATE_REPO = _(
+        'There was a problem connecting to your repository, '
+        'ensure that your repository URL is correct.'
+    )
```
Comment (ericholscher) [inline readthedocs/projects/exceptions.py]: If it's a private repo, shouldn't we warn them here? Why are these messages the same?

File: readthedocs/doc_builder/environments.py [hunk 15]
```diff
@@ -537,28 +548,40 @@ def __enter__(self):
...
+                log.exception(
+                    LOG_TEMPLATE
+                    .format(
+                        project=self.project.slug,
+                        version=self.version.slug,
+                        msg="Couldn't remove container",
+                    ),
+                )
+            self.container = None
+        except BuildEnvironmentError:
```
Comment (humitos) [inline readthedocs/doc_builder/environments.py]: I'm not following this. I think it's not possible to `BuildEnvironmentError` be raised in the `try:` block here. What could be the case?

File: readthedocs/projects/tasks.py (hunk 6 above)
Comment (ericholscher) [reply readthedocs/projects/tasks.py]: Is this logic ^^ documented somewhere?

File: readthedocs/projects/tasks.py [hunk 16]
```diff
@@ -127,24 +126,40 @@ def run(self, pk, version_pk=None, build_pk=None, record=True,
...
+                log.exception(
+                    'An unhandled exception was raised during project build',
+                    extra={'tags': {'build': build_pk}}
+                )
+                self.build_env.build['error'] = _(
+                    'An unknown error was encountered while building your project. '
+                    'Please include the build id ({build_id}) in any bug reports.'.format(
+                        build_id=build_pk
+                    ))
+                self.build_env.update_build(BUILD_STATE_FINISHED)
```
Comment (ericholscher) [inline readthedocs/projects/tasks.py]: Another place we're updating build finished..?

File: readthedocs/rtd_tests/tests/test_doc_building.py [hunk 17]
```diff
@@ -266,7 +591,23 @@ def test_container_timeout(self):
...
+            'success': False,
+            'project': self.project.pk,
+            'setup_error': u'',
+            'exit_code': 1,
+            'length': 0,
+            'error': 'Build exited due to time out',
+            'setup': u'',
+            'output': u'',
+            'state': u'finished',
+            'builder': u'foo'
```
Comment (ericholscher) [inline readthedocs/rtd_tests/tests/test_doc_building.py]: All these listings of massive argument lists seem brittle (it will break whenever we add/change/remove/ an argument). Can we just check for the actual value we care about being called, presumably `error`?

File: readthedocs/vcs_support/backends/git.py [hunk 18]
```diff
@@ -56,12 +56,9 @@ def repo_exists(self):
...
 
     def fetch(self):
-        code, _, err = self.run('git', 'fetch', '--tags', '--prune')
+        code, _, _ = self.run('git', 'fetch', '--tags', '--prune')
         if code != 0:
-            raise ProjectImportError(
-                "Failed to get code from '%s' (git fetch): %s\n\nStderr:\n\n%s\n\n" % (
-                    self.repo_url, code, err)
-            )
+            raise RepositoryError
```
Comment (ericholscher) [inline readthedocs/vcs_support/backends/git.py]: Do we want to lose the specifics of these errors?

File: readthedocs/projects/tasks.py [hunk 19]
```diff
@@ -127,24 +126,40 @@ def run(self, pk, version_pk=None, build_pk=None, record=True,
...
+            self.setup_env.build['error'] = _(
+                'An unknown error was encountered while setting up your project. '
                 'Please include the build id ({build_id}) in any bug reports.'.format(
                     build_id=build_pk
                 ))
-        finally:
-            if unhandled_failure:
-                self.build_env.build['error'] = unhandled_failure
-            self.build_env.update_build(BUILD_STATE_FINISHED)
+            self.setup_env.update_build(BUILD_STATE_FINISHED)
```
Comment (humitos) [inline readthedocs/projects/tasks.py]: Since `self.setup_env` was instantiated with `finalize=False` this won't work.

File: readthedocs/projects/tasks.py [hunk 20]
```diff
@@ -297,19 +313,10 @@ def setup_vcs(self):
         self.setup_env.update_build(state=BUILD_STATE_CLONING)
```
Comment (humitos) [inline readthedocs/projects/tasks.py]: Also a problem here. The build won't update with this status since it's `finalize=False`.

File: readthedocs/doc_builder/backends/sphinx.py (hunk 7 above)
Comment (agjohnson) [reply readthedocs/doc_builder/backends/sphinx.py]: I'm not sure either, the docs aren't clear to me:
https://pythonhosted.org/six/#six.reraise

I'll leave it, assuming someone with more knowledge of py2/3 compat did this :)

File: readthedocs/doc_builder/environments.py (hunk 10 above)
Comment (agjohnson) [reply readthedocs/doc_builder/environments.py]: Ah true. I didn't like `finalize` either. I'll rename this `commit` to match the Django usage. Would this change make more sense?

File: readthedocs/doc_builder/environments.py (hunk 8 above)
Comment (agjohnson) [reply readthedocs/doc_builder/environments.py]: The `finalize` check is in `update_build`, so it should only post if `finalize` is `True`

File: readthedocs/doc_builder/environments.py (hunk 9 above)
Comment (agjohnson) [reply readthedocs/doc_builder/environments.py]: It won't, because this is raised in `__enter__`. Errors raised in `__enter__` raise to the parent context, so we have to explicitly clean up here.

File: readthedocs/doc_builder/environments.py (hunk 15 above)
Comment (agjohnson) [reply readthedocs/doc_builder/environments.py]: `get_client` raises a `BuildEnvironmentError` on failure here. I'll rethink refactoring this as well though, perhaps a more specific exception, or allowing the docker exception to bubble up makes more sense.

File: readthedocs/projects/exceptions.py (hunk 2 above)
Comment (agjohnson) [reply readthedocs/projects/exceptions.py]: Exception is the base class, Error is the specific class that kills a build and reports to a user, Warning also is absed on Exception and doesn't kill the build.

File: readthedocs/projects/exceptions.py (hunk 14 above)
Comment (agjohnson) [reply readthedocs/projects/exceptions.py]: We aren't thoroughly testing if this was a private repo or not, we'd have to grep the git output. @humitos raised a similar question and the best answer is to execute checkout in a build environment and report the commands, not parse them to determine what the error was.

Messages aren't the same, but usage could be clearer. I should probably rename these constants, as `PRIVATE_REPO` is raised if we support private repositories -- that is, we tell the user "check your url, because it wasn't the repository privacy that caused the problem. `PUBLIC_REPO` error is "check your url or your project privacy, because we don't support private repositories."

In fact, I'll update the copy here to be more explicit about repo privacy.

File: readthedocs/projects/tasks.py (hunk 6 above)
Comment (agjohnson) [reply readthedocs/projects/tasks.py]: Yup, on the environment class. I can expand more.

File: readthedocs/projects/tasks.py (hunk 19 above)
Comment (agjohnson) [reply readthedocs/projects/tasks.py]: Ah yes, good catch!

File: readthedocs/projects/tasks.py (hunk 16 above)
Comment (agjohnson) [reply readthedocs/projects/tasks.py]: It seems this is what we want, no? This is the fallback that was added in case there are any exceptions outside of the build environment `__enter__`, normal context, or in `__exit__`, such as in `run_build`.

File: readthedocs/vcs_support/backends/git.py (hunk 18 above)
Comment (agjohnson) [reply readthedocs/vcs_support/backends/git.py]: I think surfacing an error that is helpful provides more user guidance -- that is, showing a failed checkout command isn't clear to the user that the problem is their configuration. I think the ultimate fix here is to do both and show the command execution as a `BuildCommand`. I removed this for now, for consistency on errors and because displaying the stderr output like this looked sloppy. I think we can pretty quickly move towards combining execution environments though.

File: readthedocs/doc_builder/environments.py (hunk 10 above)
Comment (humitos) [reply readthedocs/doc_builder/environments.py]: maybe the `commit` should be in the `update_build` method to avoid the other problems that I mentioned

File: readthedocs/vcs_support/backends/git.py (hunk 18 above)
Comment (humitos) [reply readthedocs/vcs_support/backends/git.py]: I think that having both in the future is better.

I like the idea of communicate the "common problem" to the user in a very simplified way, but also show the output (that could be used when debugging something and could help the user to understand it better -also if the problem is something different that we are reporting as simplified way)

File: readthedocs/rtd_tests/tests/test_doc_building.py (hunk 17 above)
Comment (humitos) [reply readthedocs/rtd_tests/tests/test_doc_building.py]: Test are failing because of this, in fact:

    'length': <ANY>

Besdies, it's complicated to find the differences when it fails. In case you want to leave as it is now, I'd suggest to use `assertDictEqual`

File: readthedocs/doc_builder/environments.py (hunk 8 above)
Comment (ericholscher) [reply readthedocs/doc_builder/environments.py]: Gah ok. Lost that in the review.

File: readthedocs/projects/tasks.py (hunk 16 above)
Comment (ericholscher) [reply readthedocs/projects/tasks.py]: Ah, ok. I got lost when reviewing this for some reason, and didn't realize where everything was getting called from.

**Comment by agjohnson** [comment]
Okay! There is even more logic now. I believe I've addressed all of the feedback and have dropped in a bunch more docs while this is fresh. One last pass would be a good idea

**Comment by humitos** [comment]
I'd like to take a look a it tomorrow

El 30 nov. 2017 7:20 p. m., "Anthony" <notifications@github.com> escribió:

Okay! There is even more logic now. I believe I've addressed all of the
feedback and have dropped in a bunch more docs while this is fresh. One
last pass would be a good idea

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/rtfd/readthedocs.org/pull/3310#issuecomment-348363713>,
or mute the thread

**Comment by agjohnson** [comment]
Pinging @humitos ☎️

File: readthedocs/doc_builder/environments.py [hunk 21]
```diff
@@ -267,23 +268,39 @@ class BuildEnvironment(object):
...
+    :py:meth:`__exit__` method, specifically, inside :py:meth:`handle_exception`
+    and :py:meth:`update_build`. If the exception is a subclass of
+    :py:cls:`BuildEnvironmentError`, then this error message is added to the
+    build object and is shown to the user as the top-level failure reason for
+    why the build failed. Other exceptions raise a general failure warning on
+    the build.
+
+    We only update the build through the API in one of three cases:
+
+    * The build is not done and needs incremental builds
```
Comment (humitos) [inline readthedocs/doc_builder/environments.py]: What is an incremental build?

File: readthedocs/doc_builder/environments.py [hunk 22]
```diff
@@ -267,23 +268,39 @@ class BuildEnvironment(object):
...
+    * The build failed and we should always report this change
+    * The build was successful and ``commit`` is ``True``
 
     :param project: Project that is being built
     :param version: Project version that is being built
     :param build: Build instance
     :param record: Record status of build object
     :param environment: shell environment variables
+    :param commit: update the build object via API if the build was successful
```
Comment (humitos) [inline readthedocs/doc_builder/environments.py]: I think this is still confusing.

When `commit=True` (the default) and we are in the middle of the build, the build won't be updated. So, I think this should have a better name (if that is possible) like `last_command`, since I think it's more representative of what it measn.

Anyway, I don't see the best solution very clear.

File: readthedocs/doc_builder/environments.py [hunk 23]
```diff
@@ -386,9 +403,12 @@ def done(self):
     def update_build(self, state=None):
         """Record a build by hitting the API
 
-        This step is skipped if we aren't recording the build, or if we don't
-        want to record successful builds yet (if we are running setup commands
-        for the build)
+        This step is skipped if we aren't recording the build. To avoid
+        recording successful builds yet (for instance, running setup commands for
+        the build), set the ``commit`` argument on environment instantiation.
```
Comment (humitos) [inline readthedocs/doc_builder/environments.py]: set the commit argument to `False`

File: readthedocs/doc_builder/environments.py [hunk 24]
```diff
@@ -417,37 +437,47 @@ def update_build(self, state=None):
...
+            # Build is done, but isn't successful, always update
+            or (self.done and not self.successful)
+            # Otherwise, are we explicitly to not update?
+            or self.commit
+        )
+        if update_build:
+            try:
+                api_v2.build(self.build['id']).put(self.build)
+            except HttpClientError as e:
+                log.error("Unable to post a new build: %s", e.content)
```
Comment (humitos) [inline readthedocs/doc_builder/environments.py]: `Unable to update the build:` and maybe add the build `id` also to the log

File: readthedocs/doc_builder/environments.py [hunk 25]
```diff
@@ -537,28 +574,43 @@ def __enter__(self):
...
+            self.update_build_from_container_state()
 
-        # Update buildenv state given any container error states first
-        self.update_build_from_container_state()
+            client = self.get_client()
+            try:
+                client.kill(self.container_id)
+            except DockerAPIError:
+                log.exception(
+                    'Unable to remove container: id=%s',
```
Comment (humitos) [inline readthedocs/doc_builder/environments.py]: `Unable to kill the container`

File: readthedocs/doc_builder/environments.py [hunk 26]
```diff
@@ -576,13 +628,21 @@ def get_client(self):
...
-                          project=self.project.slug,
-                          version=self.version.slug,
-                          msg=e),
-                      exc_info=True)
-            raise BuildEnvironmentError('Problem creating build environment')
+            log.exception(
+                LOG_TEMPLATE.format(
+                    project=self.project.slug,
+                    version=self.version.slug,
+                    msg='Could not connection to Docker API',
```
Comment (humitos) [inline readthedocs/doc_builder/environments.py]: Could not _connect_

File: readthedocs/doc_builder/environments.py [hunk 27]
```diff
@@ -655,11 +715,22 @@ def create_container(self):
...
             client.start(container=self.container_id)
+        except ConnectionError as e:
+            log.exception(
+                LOG_TEMPLATE.format(
+                    project=self.project.slug,
+                    version=self.version.slug,
+                    msg=e,
+                ),
+            )
+            raise BuildEnvironmentError('There was a problem connecting to Docker')
```
Comment (humitos) [inline readthedocs/doc_builder/environments.py]: Here, the problem is similar than https://github.com/rtfd/readthedocs.org/pull/3310/files#diff-ca52b098301dd315a834b3556ab9a7d5R638 but in this case we are communicating the Docker error to the user.

File: readthedocs/doc_builder/environments.py (hunk 22 above)
Comment (agjohnson) [reply readthedocs/doc_builder/environments.py]: probably the most explicit we can is `update_on_success`

**Comment by agjohnson** [comment]
@humitos Thanks for the thorough feedback! I'm just waiting on a final CI test here. If changes look good, lets merge!
